ocaml/typing/jkind.ml

```diff
@@ -861,86 +912,93 @@ module Jkind_desc = struct
+           (Contention.Const_op.le t.monadic_op_bounds.contention new_contention)
     in
     ( { t with
-        modes_upper_bounds =
-          { t.modes_upper_bounds with
-            portability = new_portability;
-            contention = new_contention
-          }
-      },
+        comonadic_bounds =
+          { t.comonadic_bounds with
+            portability = new_portability };
+        monadic_op_bounds =
+          { t.monadic_op_bounds with
+            contention = new_contention } },
       added_crossings )
 
   let max = of_const Const.max
 
   let equate_or_equal ~allow_mutation
       { layout = lay1;
-        modes_upper_bounds = modes1;
+        comonadic_bounds = comonadic1;
+        monadic_op_bounds = monadic1;
         externality_upper_bound = ext1;
         nullability_upper_bound = null1
       }
       { layout = lay2;
-        modes_upper_bounds = modes2;
+        comonadic_bounds = comonadic2;
+        monadic_op_bounds = monadic2;
         externality_upper_bound = ext2;
         nullability_upper_bound = null2
       } =
     Layout.equate_or_equal ~allow_mutation lay1 lay2
-    && Modes.equal modes1 modes2
+    && Comonadic_bounds.equal comonadic1 comonadic2
+    && Monadic_op_bounds.equal monadic1 monadic2
     && Externality.equal ext1 ext2
     && Nullability.equal null1 null2
 
   let sub
       { layout = lay1;
-        modes_upper_bounds = modes1;
-        externality_upper_bound = ext1;
-        nullability_upper_bound = null1
-      }
-      { layout = lay2;
-        modes_upper_bounds = modes2;
-        externality_upper_bound = ext2;
-        nullability_upper_bound = null2
-      } =
-    Misc.Le_result.combine_list
-      [ Layout.sub lay1 lay2;
-        Modes.less_or_equal modes1 modes2;
-        Externality.less_or_equal ext1 ext2;
-        Nullability.less_or_equal null1 null2 ]
-
-  let intersection
-      { layout = lay1;
-        modes_upper_bounds = modes1;
+        comonadic_bounds = comonadic1;
+        monadic_op_bounds = monadic1;
         externality_upper_bound = ext1;
         nullability_upper_bound = null1
       }
       { layout = lay2;
-        modes_upper_bounds = modes2;
+        comonadic_bounds = comonadic2;
+        monadic_op_bounds = monadic2;
         externality_upper_bound = ext2;
         nullability_upper_bound = null2
       } =
-    Option.bind (Layout.intersection lay1 lay2) (fun layout ->
-        Some
-          { layout;
-            modes_upper_bounds = Modes.meet modes1 modes2;
-            externality_upper_bound = Externality.meet ext1 ext2;
-            nullability_upper_bound = Nullability.meet null1 null2
-          })
+    let (&&<) = Misc.Le_result.combine in
+    Layout.sub lay1 lay2
+    &&< Comonadic_bounds.less_or_equal comonadic1 comonadic2
+    &&< Monadic_op_bounds.less_or_equal monadic1 monadic2
+    &&< Externality.less_or_equal ext1 ext2
+    &&< Nullability.less_or_equal null1 null2
+
+  let intersection k1 k2 =
```

Comment:
Though I like a lot of the changes you've made in this file, I liked the old version of this function more: it's both more efficient and more future-proof.

Reply:
Fair enough. Do you mind if I still swap `Option.bind` for a pattern match though? I
don’t like `bind` when it always `return`s — that should be `map` — but then `Option.map` has the parameters the other way around and that reads badly. (Also first-order >> higher-order anyway)

Reply:
Yeah, that's fine.

> Also first-order >> higher-order anyway

That's sad. I would hope we could rely on our optimizer to sort this out.

Reply:
I mean that first-order code is better than higher-order code, not that it is faster. Higher-order code is inherently harder to reason about: I have to think harder when I read that `Option.bind` code than when I read the trivial pattern match. Sometimes it is worth it to hide some amount of boiler-plate, but here the match takes about the same amount of code as the `bind` so it's a bad trade.

Reply:
I was actually thinking that the explicit pattern match makes me think harder: with an explicit match, I have to think about each case separately, whereas with `bind` or `map`, I know e.g. that `None` maps to `None`. I’m not disagreeing with you here about what’s easier for you, just observing that what’s easier depends on what abstractions you have ready-to-go in your head and what you’re used to.

To be clear, I don’t feel strongly here at all, just observing a difference. I’m happy with an explicit match.